src/main/scala/gwi/Tasks.scala

```diff
@@ -0,0 +1,52 @@
+package gwi
+
+import gwi.text_transform.StatefullTask
+
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import scala.concurrent.ExecutionContext
+import scala.jdk.CollectionConverters._
+
+case class Task[State, Result](task: StatefullTask[State, Result], future: Cancellable[Result])
+
+trait TasksList[State] {
+  def getTaskState(taskId: UUID): Option[State]
+}
+
+class Tasks[State, Result](implicit executionContext: ExecutionContext) extends TasksList[State] {
+
+  val tasks = new ConcurrentHashMap[UUID, Task[State, Result]]().asScala
+
+  def addTask(id: UUID, task: StatefullTask[State, Result]): Unit = {
+    val future = new Cancellable[Result](executionContext, task.doTask)
+    tasks.addOne(id -> Task(task, future = future))
+  }
+
+  def cancelTask(id: UUID): Unit = {
+    tasks.get(id).map(t => {
+      println(s"canceling task $id" + t)
+
+      // stop execution
+      t.task.interruptTask()
+      t.future.cancel()
```

Comment:
It seems to me that you use two interruption flags. A custom one and jvm thread built-in interruption flag.

Reply:
you are right, this part was not implemented well. i wanted to review it but had to finish due to issues at work.